### PR TITLE
Add a user-guide page on constraining an optimization problem

### LINKED ISSUE #191: document multiple ways of constraining an optimization problem in scikit-agent
PR #189 introduces the Fischer-Burmeister as a way of constraining decisions in a differentiable way with neural nets.

There's an alternative way built into some of the policy neural networks.

It would be great to have a dedicated documentation page describing the options here.

### PR DESCRIPTION
**Pull request by alanlujan91**
Closes #191. The new user-guide page documents the options for constraining a decision: declaring bounds on a `Control`, the open-bounds transforms built into the policy networks, and the Fischer-Burmeister complementarity loss from #189, including its current upper-bound-only scope and how the two mechanisms compose on the D-4 benchmark. A short closing section covers VBI's box-constraint handling so the page is not specific to neural methods.

Two small fixes ride along in `blocks.md`: its Constraints section now links to the new page, and the portfolio example passed numeric bounds that the network classes reject with `TypeError`; it now passes callables. The docs build cleanly under Sphinx.

### CODE REVIEW

File: docs/user_guide/constraints.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+constraint introduces exactly the kind of kink that gradients handle badly.
+
+scikit-agent addresses this with one declaration and two enforcement mechanisms.
+Bounds are declared once, on the {py:class}`~skagent.block.Control` object. A
+policy network can then build feasibility into its output layer, and a training
+loss can encode the optimality conditions that hold where the constraint binds.
+The two mechanisms answer different questions. The first guarantees that every
+candidate policy is feasible; the second teaches the solver where the constraint
+should bind. They compose, and for problems with binding constraints they work
+best together.
```
Comment (sbenthall) [inline docs/user_guide/constraints.md:18]: I think there are some AI-style evaluative claims, like "they work best together", that we don't have evidence for yet.

Also, this is written as if the FB constraint is fully general, though I believe it's currently implemented only on the EulerEquationLoss object?

File: docs/user_guide/constraints.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    lower_bound=lambda: 1e-3,
+    upper_bound=lambda m: m,
+)
+```
+
+For example, this declares a decision $c$ that must stay between a small
+positive floor and the pre-decision state $m$; in a consumption-saving model the
+upper bound is a borrowing constraint. Either bound may be omitted, in which
+case the control is unbounded on that side. Passing a raw number such as
+`upper_bound=1.0` raises a `TypeError` when a policy network is built from the
```
Comment (sbenthall) [inline docs/user_guide/constraints.md:38]: We should think about how we want this user interface to look ideally and make an issue for it.
This seems a bit awkard and it would be possible to handle other input types.

File: docs/user_guide/constraints.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+Fischer-Burmeister function
+
+$$
+\operatorname{FB}(f, s) = f + s - \sqrt{f^2 + s^2} = 0,
+$$
+
+which is zero exactly where the complementarity conditions hold and
+differentiable everywhere. Setting `constrained=True` on
+{py:class}`~skagent.loss.EulerEquationLoss` builds this residual for every
+control that declares an `upper_bound`:
```
Comment (sbenthall) [inline docs/user_guide/constraints.md:91]: is there any other way to use it?

File: docs/user_guide/constraints.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+explicit `upper_bound` falls back to the one-sided penalty
+$\operatorname{relu}(-f_a) \cdot \operatorname{relu}(-f_b)$, which punishes only
+violations of $f \geq 0$.
+
+One scope restriction is worth stating plainly: the constrained loss currently
+models the upper-bound side of the complementarity condition only. A
+`lower_bound` declared on a `Control` is respected by the network transform
+above, but it does not yet enter the complementarity residual; bilateral support
+requires flipping the residual sign for lower-binding cases
+($\operatorname{FB}(-f, x - \underline{x})$) and is left as a follow-up.
```
Comment (sbenthall) [inline docs/user_guide/constraints.md:109]: I don't totally understand what this means, and I expect the user reading the 'user guide' won't either.

It looks like something that should be implemented hasn't been implemented yet.

I wonder if it would be possible to write a test that exposes this issue, and fix it, removing the need for this paragraph.

File: docs/user_guide/constraints.md [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+## Using Both Together
+
+The mechanisms are complementary rather than competing. The bounded network
+keeps every training iterate feasible; the Fischer-Burmeister loss supplies the
+optimality condition that tells the solver where the bound binds and by how
+much. The constrained perfect-foresight benchmark (`D-4` in the
+{doc}`benchmark_models` registry) is the in-package demonstration: a policy
+network trained on `EulerEquationLoss(constrained=True)` matches a
```
Comment (sbenthall) [inline docs/user_guide/constraints.md:118]: It would make sense to demonstrate this with a code snippet here.

File: docs/user_guide/constraints.md [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+A practical rule of thumb: if the constraint cannot bind in the region of the
+state space you care about, the default network bounds are enough. If it can
+bind and you train on an Euler or first-order-condition objective, add
+`constrained=True` so the loss knows what optimality looks like at the boundary.
+
+## Grid Solvers
+
+Constraint handling is not specific to neural methods. The value backwards
+induction solver ({py:mod}`skagent.algos.vbi`) reads the same `Control` bounds
+and passes them to `scipy.optimize.minimize` as box constraints on each grid
```
Comment (sbenthall) [inline docs/user_guide/constraints.md:134]: I wonder if this is expose too much about the underlying code for the 'user guide'.
I feel like the 'user guide' should walk the user through how to use the library.

In short, I think this guide could be simplified, but it needs to concisely state the scope of when the options for constraints are available.
